File: backend/app/graphql/mutations/register_user.rb
module Mutations
  class RegisterUser < Mutations::BaseMutation
    graphql_name "RegisterUser"

    argument :first_name, String, required: true
    argument :last_name, String, required: true
    argument :alias_name, String, required: true
    argument :email, String, required: true
    argument :password, String, required: true

    field :user, Types::UserType, null: false

    def resolve(args)
      begin
        user = User.create!(
          first_name: args[:first_name],
          last_name: args[:last_name],
          alias_name: args[:alias_name],
          email: args[:email],
          password: args[:password],
        )

        MutationResult.call(
          obj: { user: user },
          success: user.persisted?,
          errors: user.errors,
        )
      rescue ActiveRecord::RecordInvalid => exception
        GraphQL::ExecutionError.new(exception.message)
      end
    end
  end
end

File: frontend/src/mutations/registerUser.js
import gql from 'graphql-tag';

const mutation = gql`
  mutation registerUser(
    $firstName: String!,
    $lastName: String!,
    $aliasName: String!,
    $email: String!,
    $password: String!,
  ) {
    registerUser(input: {
      firstName: $firstName,
      lastName: $lastName,
      aliasName: $aliasName,
      email: $email,
      password: $password,
    }) {
      user {
        id
        firstName
        lastName
        aliasName
        email
        authenticationToken
      }
      success
      errors
    }
  }
`;

export default function({
  apollo,
  firstName,
  lastName,
  aliasName,
  email,
  password,
}) {
  return apollo.mutate({
    mutation,
    variables: {
      firstName,
      lastName,
      aliasName,
      email,
      password,
    },
  });
}